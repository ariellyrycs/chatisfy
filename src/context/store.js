import React, { useReducer, createContext, useContext } from 'react';
import Mock from '../Mock';
import { generateUId, generateTime } from '../Utils/General.js';


const currentUser = {
	id: generateUId(),
	nickName: ''
};

const Context = createContext();

const useAppContext = () => useContext(Context);

const initialState = {
	...Mock,
	currentUser
};

const mainReduce = (state, action) => {
	switch (action.type) {
		case 'UPDATE_ROOM':
			return {
				...state,
				selectedRoom: action.roomId
			};
		case 'SEND_MESSAGE':
			let currentId = generateUId();
			let rooms = state.rooms;
			let messageByRooms = rooms.messages;
			let messages = state.messages;
			let currentRoomId = state.selectedRoom,
				sentBy = action.sentBy;
			return {
				...state,
				rooms: {
					...rooms,
					messages: {
						...rooms.messages,
						[currentRoomId]: [...(messageByRooms[currentRoomId] || []), currentId]
					}
				},
				messages: {
					...messages,
					data: {
						...messages.data,
						[currentId]: {
							id: currentId,
							from: sentBy,
							content: action.message,
							createdTime: generateTime(),
							seenBy: [],
							own: action.from === sentBy
						}
					}
				}
			};
		case 'UPDATE_NICKNAME':
			return {
				...state,
				currentUser: {
					...state.currentUser,
					nickName: action.nickName
				}
			};
		default:
			return state;
	}
};

function ContextProvider(props) {
	const [state, dispatch] = useReducer(mainReduce, initialState);
	return (
		<Context.Provider value={{ state, dispatch }}>
			{props.children}
		</Context.Provider>
	);
};

export { ContextProvider, Context, useAppContext };