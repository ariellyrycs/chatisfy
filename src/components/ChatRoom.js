import React, { useState } from 'react';
import { useAppContext } from '../context/store';
import Messages from './Messages';
import { Grid, Typography, TextField, Button, Box } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import clsx from 'clsx';

const useStyles = makeStyles({
    spacingAroundSm: {
        margin: '10px 0'
    },
    staticContainer: {
        position: 'relative'
    }
});

export default function ChatRoom() {

    const [message, setMessage] = useState('');

    const { state, dispatch } = useAppContext();

    const classes = useStyles();

    const currentRoom = state.rooms.data[state.selectedRoom];

    const currentMessagesIds = state.rooms.messages[state.selectedRoom];

    const sendMessage = (e) => {
        e.preventDefault();
        if (message.trim() === '') return;
        //TODO: make call to save use effect
        dispatch({ type: 'SEND_MESSAGE', message, from: state.currentUser.currentUserId });

        setMessage('');
        return false;
    };

    const handleEnter = e => {
        if (e.keyCode === 13 && !e.shiftKey) {
            sendMessage(e);
        }
        return false;
    };

    return (
        <Grid container direction='column'>
            <Grid item className={clsx(classes.spacingAroundSm, classes.staticContainer)}>
                <Box textAlign='center'>
                    <Typography variant='h6' component='h2'>{currentRoom.name}</Typography>
                </Box>
                <Messages currentMessagesIds={currentMessagesIds} />
            </Grid>
            <Grid item className={classes.spacingAroundSm}>
                <form onSubmit={sendMessage}>
                    <Grid container justify='center'>
                        <Grid item xs={8}>
                            <TextField
                                id='message'
                                label='Write here your message'
                                multiline
                                rowsMax={2}
                                value={message}
                                onChange={e => setMessage(e.target.value)}
                                onKeyDown={handleEnter}
                                fullWidth={true}
                                inputRef={input => input && input.focus()}
                            />
                        </Grid>
                        <Grid item md={3} className={classes.spacingAroundSm}>
                            <Button
                                type='submit'
                                variant='contained'
                                color='primary'
                                fullWidth={true}>
                                Send
                            </Button>
                        </Grid>
                    </Grid>
                </form>
            </Grid>
        </Grid>
    );
}