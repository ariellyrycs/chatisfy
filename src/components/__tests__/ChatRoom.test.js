import React from 'react';
import * as Store from '../../context/store';
import ChatRoom from '../ChatRoom';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json'


describe('<ChatRoom />', () => {
    test('should render chat room component', () => {
        const contextValues = {
            state: {
                rooms: {
                    data: {
                        1: {
                            name: 'React is awesome Group',
                            thumbnail: ''
                        }
                    },
                    messages: {
                        1: [1,2,3,4]
                    }
                },
                
                selectedRoom: 1
            },
            dispatch: () => {}
        };
        jest.spyOn(Store, 'useAppContext')
            .mockImplementation(() => contextValues);
        const tree = shallow(<ChatRoom />);
        expect(toJson(tree)).toMatchSnapshot();
    });
});