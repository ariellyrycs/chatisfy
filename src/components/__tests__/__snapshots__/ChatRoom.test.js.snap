// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatRoom /> should render chat room component 1`] = `
<WithStyles(ForwardRef(Grid))
  container={true}
  direction="column"
>
  <WithStyles(ForwardRef(Grid))
    className="makeStyles-spacingAroundSm-1 makeStyles-staticContainer-2"
    item={true}
  >
    <Styled(MuiBox)
      textAlign="center"
    >
      <WithStyles(ForwardRef(Typography))
        component="h2"
        variant="h6"
      >
        React is awesome Group
      </WithStyles(ForwardRef(Typography))>
    </Styled(MuiBox)>
    <Messages
      currentMessagesIds={
        Array [
          1,
          2,
          3,
          4,
        ]
      }
    />
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    className="makeStyles-spacingAroundSm-1"
    item={true}
  >
    <form
      onSubmit={[Function]}
    >
      <WithStyles(ForwardRef(Grid))
        container={true}
        justify="center"
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
          xs={8}
        >
          <WithStyles(ForwardRef(TextField))
            fullWidth={true}
            id="message"
            inputRef={[Function]}
            label="Write here your message"
            multiline={true}
            onChange={[Function]}
            onKeyDown={[Function]}
            rowsMax={2}
            value=""
          />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          className="makeStyles-spacingAroundSm-1"
          item={true}
          md={3}
        >
          <WithStyles(ForwardRef(Button))
            color="primary"
            fullWidth={true}
            type="submit"
            variant="contained"
          >
            Send
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </form>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
